common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -139,9 +139,55 @@ public static UTF8String fromAddress(Object base, long offset, int numBytes) {
    * Creates an UTF8String from String.
    */
   public static UTF8String fromString(String str) {
+    if (str == null) {
+      return null;
+    }
+    byte[] bytes;
+    // Optimization for ASCII strings: construct an exactly-right-sized
+    // array ourselves. CharsetEncoder ends up over-allocating for ASCII
+    // strings, leading to extra memory copies and garbage creation.
+    if (isAscii(str)) {
+      bytes = new byte[str.length()];
+      for (int i = 0; i < str.length(); i++) {
+        bytes[i] = (byte) str.charAt(i);
+      }
+    } else {
+      bytes = str.getBytes(StandardCharsets.UTF_8);
+    }
+    return fromBytes(bytes);
+  }
+
+  /**
+   * Pre-SPARK-27841 version of fromString(), retained for benchmarking purposes.
+   */
+  public static UTF8String fromStringSlow(String str) {
     return str == null ? null : fromBytes(str.getBytes(StandardCharsets.UTF_8));
   }
 
+  /**
+   * Returns true if the given string consists entirely of ASCII characters, false otherwise.
+   */
+  private static boolean isAscii(String str) {
+    for (int i = 0; i < str.length(); i++) {
+      if (str.charAt(i) > 127) { // unsigned comparison
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Returns true if the given UTF8 bytes consist entirely of ASCII characters, false otherwise.
+   */
+  private static boolean isAscii(byte[] bytes) {
```

Comment:
It would be interesting to benchmark your implementation against to:
```scala
private static boolean isAscii(byte[] bytes) {
    int length = bytes.length;
    int s = 0;
    for (int i = 0; i < length; i++) {
      s |= bytes[i] & 0x80;
    }
    return s == 0;
  }
```
It doesn't have the problem of branch mispredictions. Maybe it will be faster in general case.

Reply:
Good idea! I think we can go one step further and move the `& 0x80` out of the loop body and into the final `return` statement, shaving a couple of instructions out of the loop body.

Just pushed a change implementing this.

Reply:
I think that the HW branch predictor can reduce misprediction since the direction is mostly the same. If we assume that ASCII case is dominant, the compiler optimizer may apply simdization.

If this code is faster than the original one, how about applying this to `isAscii(String str)`?